src/backend/distributed/executor/insert_select_executor.c

```diff
@@ -0,0 +1,150 @@
+ * distributed table and then executing the SELECT query using that DestReceiver
+ * as the tuple destination.
+ */
+TupleTableSlot *
+CoordinatorInsertSelectExecScan(CustomScanState *node)
+{
+	CitusScanState *scanState = (CitusScanState *) node;
+	TupleTableSlot *resultSlot = NULL;
+
+	if (!scanState->finishedRemoteScan)
+	{
+		EState *executorState = scanState->customScanState.ss.ps.state;
+		MultiPlan *multiPlan = scanState->multiPlan;
+		Query *selectQuery = multiPlan->insertSelectQuery;
+		List *insertTargetList = multiPlan->insertTargetList;
+		Oid targetRelationId = multiPlan->targetRelationId;
+
+		ereport(DEBUG1, (errmsg("Collecting INSERT ... SELECT results on coordinator")));
+
+		ExecuteSelectIntoRelation(targetRelationId, insertTargetList, selectQuery,
+								  executorState);
+
+		scanState->finishedRemoteScan = true;
+	}
+
+	resultSlot = ReturnTupleFromTuplestore(scanState);
+
+	return resultSlot;
+}
+
+
+/*
+ * ExecuteSelectIntoRelation executes given SELECT query and inserts the
+ * results into the target relation, which is assumed to be a distributed
+ * table.
+ */
+static void
+ExecuteSelectIntoRelation(Oid targetRelationId, List *insertTargetList,
+						  Query *selectQuery, EState *executorState)
+{
+	ParamListInfo paramListInfo = executorState->es_param_list_info;
+
+	ListCell *insertTargetCell = NULL;
+	List *columnNameList = NIL;
+	bool stopOnFailure = false;
+
+	CitusCopyDestReceiver *copyDest = NULL;
+
+	BeginOrContinueCoordinatedTransaction();
+
+	/* build the list of column names for the COPY statement */
+	foreach(insertTargetCell, insertTargetList)
+	{
+		TargetEntry *insertTargetEntry = (TargetEntry *) lfirst(insertTargetCell);
+
+		columnNameList = lappend(columnNameList, insertTargetEntry->resname);
+	}
+
+	/* set up a DestReceiver that copies into the distributed table */
+	copyDest = CreateCitusCopyDestReceiver(targetRelationId, columnNameList,
+										   executorState, stopOnFailure);
+
+	ExecuteIntoDestReceiver(selectQuery, paramListInfo, (DestReceiver *) copyDest);
+
+	executorState->es_processed = copyDest->tuplesSent;
+
+	XactModificationLevel = XACT_MODIFICATION_DATA;
+}
+
+
+/*
+ * ExecuteIntoDestReceiver plans and executes a query and sends results to the given
+ * DestReceiver.
+ */
+static void
+ExecuteIntoDestReceiver(Query *query, ParamListInfo params, DestReceiver *dest)
+{
+	PlannedStmt *queryPlan = NULL;
+	Portal portal = NULL;
+	int eflags = 0;
+	int cursorOptions = 0;
+	long count = FETCH_ALL;
+
+	/* create a new portal for executing the query */
+	portal = CreateNewPortal();
+
+	/* don't display the portal in pg_cursors, it is for internal use only */
+	portal->visible = false;
+
+#if (PG_VERSION_NUM >= 90600)
+	cursorOptions = CURSOR_OPT_PARALLEL_OK;
+#endif
+
+	/* plan the subquery, this may be another distributed query */
+	queryPlan = pg_plan_query(query, cursorOptions, params);
+
+	PortalDefineQuery(portal,
```

Comment:
We can't perform the subquery in a separate session since we wouldn't see uncommitted writes and the subquery may perform modifications that we don't want to commit immediately. Note that SPI also doesn't start a separate session, it's essentially just a wrapper around the `Portal` functions.

I can't seem to reproduce that error, but there will be transactions which cannot be executed due to the way we assign connections, similar to COPY. Over time we can hopefully reduce these.